android/KMEA/app/src/main/java/com/tavultesoft/kmea/packages/LexicalModelPackageProcessor.java

```diff
@@ -0,0 +1,228 @@
+        models[i].put(KMManager.KMKey_LanguageID, languages.getJSONObject(i).getString("id"));
+        models[i].put(KMManager.KMKey_LanguageName, languages.getJSONObject(i).getString("name"));
+
+        if (welcomeExists(packageId)) {
+          File kmpFile = new File(packageId + ".kmp");
+          if (!kmpFile.exists()) {
+            kmpFile = new File(packageId + ".model.kmp");
+          }
+          File packageDir = constructPath(kmpFile, false);
+          File welcomeFile = new File(packageDir, "welcome.htm");
+          // Only storing relative instead of absolute paths as a convenience for unit tests.
+          models[i].put(KMManager.KMKey_CustomHelpLink, welcomeFile.getPath());
+        }
+      }
+      return models;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * The master KMP processing method; use after a .model.kmp download to fully install within the filesystem.
+   * @param path Filepath of a newly downloaded .kmp file.
+   * @return A list of data maps of the newly installed and/or newly upgraded lexical models found in the package.
+   * May be empty if the package file is actually an old version.
+   * <br/><br/>
+   * @throws IOException
+   * @throws JSONException
+   */
+  public static List<Map<String, String>> processKMP(File path) throws IOException, JSONException {
+    return processKMP(path, false);
+  }
+
+  /**
+   * The master KMP processing method; use after a .model.kmp download to fully install within the filesystem.
+   * @param path Filepath of a newly downloaded .kmp file.
+   * @param force A <code><b>true</b></code> value indicates that attempts to downgrade should proceed/be forced.
+   * @return A list of data maps of the newly installed and/or newly upgraded lexical models found in the package.
+   * May be empty if the package file is actually an old version.
+   * <br/><br/>
+   * @throws IOException
+   * @throws JSONException
+   */
+  public static List<Map<String, String>> processKMP(File path, boolean force) throws IOException, JSONException {
+    return processKMP(path, force, false);
+  }
+
+  /**
+   * The master KMP processing method; use after a .model.kmp download to fully install within the filesystem.
+   * @param path Filepath of a newly downloaded .kmp file.
+   * @param force A <code><b>true</b></code> value indicates that attempts to downgrade should proceed/be forced.
+   * @param preExtracted Indicates that the specified file has already been unzipped.  Only of use for testing.
+   * @return A list of data maps of the newly installed and/or newly upgraded lexical models found in the package.
+   * May be empty if the package file is actually an old version.
+   * <br/><br/>
+   * @throws IOException
+   * @throws JSONException
+   */
+  static List<Map<String, String>> processKMP(File path, boolean force, boolean preExtracted) throws IOException, JSONException {
+    // Block reserved namespaces, like /cloud/.
```

Comment:
This _really_ should not be a fully-duplicated method of the one from `PackageProcessor` - there's a ton of common code.

A little bit of work and you should be able to abstract the original to function as a common helper, allowing each class's `public` version to simply pass it the appropriate constants/callbacks for the check - and that's even if you leave all the methods static.

Alternatively, redesigning the classes to be instantiated allows use of polymorphism to greatly streamline everything here with the abstraction it facilitates.

Either way, let's try to avoid adding technical debt by keeping most of the maintenance to a single, common location rather than requiring duplicated work in the future.

Reply:
k, I've done a round of refactoring so PackageProcessor can be instantiated, and LexicalModelPackageProcessor extends as needed.